max/python/max/pipelines/architectures/gemma3multimodal/model.py

```diff
@@ -171,22 +314,517 @@ def estimate_kv_cache_size(
+        ]
+
+        return Gemma3MultiModalModelInputs(
+            tokens=next_tokens,
+            input_row_offsets=next_row_offsets,
+            return_n_logits=prev_model_inputs.return_n_logits,
+            signal_buffers=self.signal_buffers,
+            kv_cache_inputs=prev_model_inputs.kv_cache_inputs,
+            pixel_values=None,
+        )
+
+    def _prepare_vision_inputs(
+        self, context_batch: Sequence[TextAndVisionContext]
+    ) -> list[Tensor] | None:
+        """Use the VisionStacker to prepare batched vision inputs from multiple contexts.
+        The Tokenizer should have already processed images into pixel_values (pan and scan etc)"""
+        images = []
+        for context in context_batch:
+            for img in context.next_images:
+                images.append(img.pixel_values)
+
+        if not images:
+            return None
+
+        final_images = _VisionStacker().stack(images)
+
+        tensor = Tensor.from_numpy(final_images)
+        tensor = cast_dlpack_to(
+            final_images, DType.float32, DType.bfloat16, self.devices[0]
+        )
+
+        return [tensor.to(dev) for dev in self.devices]
+
+    def _batch_image_token_indices(
+        self, context_batch: Sequence[TextAndVisionContext]
+    ) -> list[Tensor] | None:
+        """Batch image token indices from multiple contexts, adjusting for
+        position in batch.
+        """
+        indices_and_offsets = []
+        batch_offset = 0
+
+        for ctx in context_batch:
+            input_ids = ctx.next_tokens
+
+            # Find where image tokens appear
+            special_image_token_mask = (
+                input_ids == self.config.image_token_index
+            )
+            indices = np.where(special_image_token_mask)[0]
+
+            if len(indices) > 0:
+                indices_and_offsets.append(indices + batch_offset)
+
+            batch_offset += ctx.active_length
+
+        if not indices_and_offsets:
+            return [
+                Tensor.zeros(shape=[0], dtype=DType.int32).to(self.devices[0])
+            ]
+
+        np_indices = np.concatenate(indices_and_offsets).astype(
+            np.int32, copy=False
+        )
+
+        # Create tensor and distribute to device
+        return [Tensor.from_numpy(np_indices).to(dev) for dev in self.devices]
+
+    def load_kv_manager(
+        self, session: InferenceSession, available_cache_memory: int | None
+    ) -> PagedKVCacheManager:
+        """Loads and initializes the KVCacheManager for the Gemma3 model.
+
+        Configures the KV cache manager based on model parameters, pipeline settings,
+        and available memory.
+
+        Args:
+            session: The MAX Engine inference session.
+            available_cache_memory: The amount of memory available for the KV cache in bytes.
+
+        Returns:
+            An initialized :obj:`KVCacheManager` instance.
+        """
+        return load_kv_manager(
```

Comment:
```suggestion
        if available_cache_memory is not None:
            available_cache_memory = int(available_cache_memory * 0.9)

        return load_kv_manager(
```
Could you add this workaround? Otherwise our memory estimation fails, and we OOM